File: node/examples/isolated-service.ts
import { RPCClass } from '../src/decorators.js';
import { createRPCClient } from '../src/index.js';

@RPCClass
class ComputeService {
  async fibonacci(n: number): Promise<number> {
    console.log(`Computing fibonacci(${n})`);
    if (n <= 1) return n;

    let a = 0,
      b = 1;
    for (let i = 2; i <= n; i++) {
      const temp = a + b;
      a = b;
      b = temp;
    }

    await new Promise((resolve) => setTimeout(resolve, 100));
    return b;
  }

  async *generatePrimes(limit: number) {
    console.log(`Generating primes up to ${limit}`);

    for (let num = 2; num <= limit; num++) {
      let isPrime = true;

      for (let i = 2; i * i <= num; i++) {
        if (num % i === 0) {
          isPrime = false;
          break;
        }
      }

      if (isPrime) {
        yield num;
        await new Promise((resolve) => setTimeout(resolve, 50));
      }
    }
  }
}

@RPCClass
class EchoService {
  async echo(message: string): Promise<string> {
    console.log(`Echo: ${message}`);
    return `Echo: ${message}`;
  }

  async ping(): Promise<string> {
    return 'pong';
  }
}

async function main() {
  const totalStart = performance.now();
  console.log('Starting isolated service connection test...\n');

  const server = createRPCClient({
    servers: ['nats://localhost:4222'],
    auth: { user: 'server', password: 'server_password' },
    name: 'service-server',
  });

  await server.connect();
  console.log('Server connected');

  const service1 = await server.service.registerHandler(
    {
      name: 'compute',
      version: '1.0.0',
      description: 'Heavy computation service',
    },
    new ComputeService(),
  );

  const service2 = await server.service.registerHandler(
    {
      name: 'echo',
      version: '1.0.0',
      description: 'Lightweight echo service',
    },
    new EchoService(),
  );

  console.log('Services registered\n');

  const client = createRPCClient({
    servers: ['nats://localhost:4222'],
    auth: { user: 'server', password: 'server_password' },
    name: 'test-client',
  });

  await client.connect();
  console.log('Client connected\n');

  console.log('Creating Service Proxies');

  // prettier-ignore
  const compute = await client.createServiceProxy<{
    fibonacci(n: number): Promise<number>;
    generatePrimes(limit: number): AsyncGenerator<number>;
  }>('compute', {
    isolatedConnection: true,
    timeout: 60000, // 60 seconds for heavy computations
  });
  console.log('Compute service proxy created (isolated connection)');

  const echo = await client.createServiceProxy<{
    echo(message: string): Promise<string>;
    ping(): Promise<string>;
  }>('echo');
  console.log('Echo service proxy created (shared connection)\n');

  console.log('Testing Concurrent Operations');

  console.log('Starting heavy computation (fibonacci)...');
  const fibStart = performance.now();
  const fibPromise = compute.proxy.fibonacci(40).then((result) => {
    const elapsed = performance.now() - fibStart;
    console.log(`Fibonacci(40) = ${result} (${elapsed.toFixed(1)}ms)`);
    return result;
  });

  console.log('\nTesting echo service while computing...');
  const echoTimes: number[] = [];
  for (let i = 1; i <= 5; i++) {
    const start = performance.now();
    const result = await echo.echo(`Message ${i}`);
    const elapsed = performance.now() - start;
    echoTimes.push(elapsed);
    console.log(`${result} (${elapsed.toFixed(1)}ms)`);
    await new Promise((resolve) => setTimeout(resolve, 200));
  }
  const avgEchoTime = echoTimes.reduce((a, b) => a + b, 0) / echoTimes.length;
  console.log(`Average echo time: ${avgEchoTime.toFixed(1)}ms`);

  console.log('\nWaiting for fibonacci computation...');
  await fibPromise;

  console.log('\nTesting Streaming with Isolation');
  const streamStart = performance.now();
  const primes = compute.proxy.generatePrimes(30);
  const collectedPrimes: number[] = [];

  console.log('Collecting primes while doing other work...');
  const primePromise = (async () => {
    for await (const prime of primes) {
      collectedPrimes.push(prime);
      console.log(`Received prime: ${prime}`);
    }
  })();

  for (let i = 1; i <= 3; i++) {
    await new Promise((resolve) => setTimeout(resolve, 100));
    const pong = await echo.ping();
    console.log(`Ping ${i}: ${pong}`);
  }

  await primePromise;
  const streamTime = performance.now() - streamStart;
  console.log(`\nCollected ${collectedPrimes.length} primes in ${streamTime.toFixed(1)}ms: ${collectedPrimes.join(', ')}`);

  console.log('\nTesting Connection Isolation');
  console.log('Disconnecting compute service (isolated)...');
  await compute.close();
  console.log('Isolated connection closed\n');

  console.log('Testing echo service after compute disconnect...');
  try {
    const start = performance.now();
    const result = await echo.echo('Still working?');
    const elapsed = performance.now() - start;
    console.log(`Echo service: ${result} (${elapsed.toFixed(1)}ms)`);
  } catch (error: any) {
    console.error('Echo failed:', error.message);
  }

  console.log('\nTesting compute service after disconnect...');
  try {
    await compute.proxy.fibonacci(10);
  } catch (error: any) {
    console.log('Expected error:', error.message);
  }

  console.log('\nCleaning up...');
  await service1.stop();
  await service2.stop();
  await client.disconnect();
  await server.disconnect();

  const totalElapsed = performance.now() - totalStart;
  console.log(`\nAll tests completed! Total time: ${totalElapsed.toFixed(1)}ms`);
}

main().catch(console.error);

File: python/examples/isolated_service.py
import asyncio
import time
from collections.abc import AsyncGenerator

import uvloop

from camera_ui_rpc import RPCClass, ServiceConfig, create_rpc_client


@RPCClass
class ComputeService:
    async def fibonacci(self, n: int) -> int:
        print(f"Computing fibonacci({n})")
        if n <= 1:
            return n

        a, b = 0, 1
        for _ in range(2, n + 1):
            a, b = b, a + b

        await asyncio.sleep(0.1)
        return b

    async def generate_primes(self, limit: int) -> AsyncGenerator[int, None]:
        print(f"Generating primes up to {limit}")

        for num in range(2, limit + 1):
            is_prime = True

            for i in range(2, int(num**0.5) + 1):
                if num % i == 0:
                    is_prime = False
                    break

            if is_prime:
                yield num
                await asyncio.sleep(0.05)


@RPCClass
class EchoService:
    async def echo(self, message: str) -> str:
        print(f"Echo: {message}")
        return f"Echo: {message}"

    async def ping(self) -> str:
        return "pong"


async def main():
    total_start = time.perf_counter()
    print("Starting isolated service connection test...\n")

    server = create_rpc_client(
        {
            "servers": ["nats://localhost:4222"],
            "auth": {"user": "server", "password": "server_password"},
            "name": "service-server",
        }
    )

    await server.connect()
    print("Server connected")

    compute_service = await server.service.register_handler(
        ServiceConfig(
            name="compute",
            version="1.0.0",
            description="Heavy computation service",
        ),
        ComputeService(),
    )

    echo_service = await server.service.register_handler(
        ServiceConfig(
            name="echo",
            version="1.0.0",
            description="Lightweight echo service",
        ),
        EchoService(),
    )

    print("Services registered\n")

    client = create_rpc_client(
        {
            "servers": ["nats://localhost:4222"],
            "auth": {"user": "server", "password": "server_password"},
            "name": "test-client",
        }
    )

    await client.connect()
    print("Client connected\n")

    print("--- Creating Service Proxies ---")

    compute_proxy = await client.create_service_proxy(
        "compute",
        isolated_connection=True,
        timeout=60000,  # 60 seconds for heavy computations
    )
    compute_proxy_service: ComputeService = compute_proxy.proxy
    print("Compute service proxy created (isolated connection)")

    echo_proxy_service = await client.create_service_proxy("echo", EchoService, isolated_connection=False)
    print("Echo service proxy created (shared connection)\n")

    print("--- Testing Concurrent Operations ---")

    print("Starting heavy computation (fibonacci)...")
    fib_start = time.perf_counter()

    async def compute_fibonacci() -> int:
        result = await compute_proxy_service.fibonacci(40)
        elapsed = (time.perf_counter() - fib_start) * 1000
        print(f"Fibonacci(40) = {result} ({elapsed:.1f}ms)")
        return result

    fib_task = asyncio.create_task(compute_fibonacci())

    print("\nTesting echo service while computing...")
    echo_times: list[float] = []
    for i in range(1, 6):
        start = time.perf_counter()
        result = await echo_proxy_service.echo(f"Message {i}")
        elapsed = (time.perf_counter() - start) * 1000
        echo_times.append(elapsed)
        print(f"{result} ({elapsed:.1f}ms)")
        await asyncio.sleep(0.2)
    avg_echo_time = sum(echo_times) / len(echo_times)
    print(f"Average echo time: {avg_echo_time:.1f}ms")

    print("\nWaiting for fibonacci computation...")
    await fib_task

    print("\n--- Testing Streaming with Isolation ---")
    stream_start = time.perf_counter()
    primes_gen = compute_proxy_service.generate_primes(30)
    collected_primes: list[int] = []

    print("Collecting primes while doing other work...")

    async def collect_primes() -> None:
        async for prime in primes_gen:
            collected_primes.append(prime)
            print(f"Received prime: {prime}")

    prime_task = asyncio.create_task(collect_primes())

    for i in range(1, 4):
        await asyncio.sleep(0.1)
        pong = await echo_proxy_service.ping()
        print(f"Ping {i}: {pong}")

    await prime_task
    stream_time = (time.perf_counter() - stream_start) * 1000
    print(
        f"\nCollected {len(collected_primes)} primes in {stream_time:.1f}ms: {', '.join(map(str, collected_primes))}"
    )

    print("\n--- Testing Connection Isolation ---")
    print("Disconnecting compute service (isolated connection)...")
    await compute_proxy.close()
    print("Isolated connection closed")

    print("\nTesting echo service after compute disconnect...")
    try:
        start = time.perf_counter()
        result = await echo_proxy_service.echo("Still working?")
        elapsed = (time.perf_counter() - start) * 1000
        print(f"Echo service: {result} ({elapsed:.1f}ms)")
    except Exception as e:
        print(f"Echo failed: {e}")

    print("\nTesting compute service after disconnect...")
    try:
        await compute_proxy_service.fibonacci(10)
        print("ERROR: Compute service should have failed!")
    except Exception as e:
        print(f"Expected error: {e}")

    print("\nCleaning up...")
    await compute_service.stop()
    await echo_service.stop()
    await client.disconnect()
    await server.disconnect()

    total_elapsed = (time.perf_counter() - total_start) * 1000
    print(f"\nTest completed! Total time: {total_elapsed:.1f}ms")


if __name__ == "__main__":
    try:
        uvloop.run(main())
    except KeyboardInterrupt:
        print("\nTest interrupted by user")
    except Exception as e:
        print(f"Error in native channel request test: {e}")
        import traceback

        traceback.print_exc()
